### PR TITLE
Histogram module

### LINKED ISSUE #23: Create a module that displays histogram.
NOTE: This may end up in 'visualize R object' module.

### PR DESCRIPTION
**Pull request by mateuszwyszynski**
This aims to resolve #23 

For now this module creates a specific histogram as in Time Since Fire part of LandWeb app

### CODE REVIEW

File: R/histogramForRaster.R [hunk 1]
```diff
@@ -0,0 +1,35 @@
+#' Histogram for Raster Shiny Module
+#'
+#' @description This creates UI for shiny module creating histogram of data contained in raste
+#'
+#' @param id An ID string that corresponds with the ID used to call the module server function
+#'
+#' @export
```
Comment (achubaty) [inline R/histogramForRaster.R]: please be sure to explicitly state any imports from other packages, and ensure all imported packages are listed in `DESCRIPTION`:

```r
#' @importFrom shinycssloaders withSpinner
#' @importFrom shiny h4 NS
#' @importFrom shinydashboard box
```

Comment (achubaty) [reply R/histogramForRaster.R]: need to add `shinycssloaders` to Imports in DESCRIPTION

Comment (achubaty) [reply R/histogramForRaster.R]: also, please add an explict `@rdname` tag to the documentation. We've been putting both ui+server portions of each shiny module in the same help file. *I.e.*:

```r
#' @rdname histogramForRaster
histogramForRasterUI <- function(id) {...}
```

and

```r
#' @rdname histogramForRaster
histogramForRaster <- function(input, output, session, raster) {...}
```

File: R/histogramForRaster.R [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+#' Histogram for Raster Shiny Module
+#'
+#' @description
+#'
+#' @param input Shiny server input object
+#' @param output Shiny server output object
+#' @param session Shiny server session object
+#' @param raster Reactive value containing raster
+#'
+#' @export
```
Comment (achubaty) [inline R/histogramForRaster.R]: add imports and enusre necessary packages are added to DESCRIPTION:

```r
#' @importFrom graphics axis barplot hist
#' @importFrom raster maxValue
#' @importFrom shiny renderPlot
```

File: R/histogramForRaster.R (hunk 1 above)
Comment (DamianRodziewicz) [reply R/histogramForRaster.R]: Added separate task to set rdnames for other modules as well. https://github.com/PredictiveEcology/SpaDES.shiny/issues/46

**Comment by DamianRodziewicz** [comment]
@mateuszwyszynski please address Alex comments. Other than that LGTM - we'll make it more generic while working on other modules as well.

**Comment by mateuszwyszynski** [comment]
Comments addressed, ptal

File: R/histogramForRaster.R [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+#' @param raster Reactive value containing raster
+#'
+#' @importFrom shiny renderPlot
+#' @importFrom raster maxValue hist
+#' @importFrom graphics axis barplot
+#'
+#' @rdname histogramForRaster
+#'
+#' @export
+histogramForRaster <- function(input, output, session, raster) {
```
Comment (DamianRodziewicz) [inline R/histogramForRaster.R]: Please revisit this module and extract all the mystery guests from this method. I can see `rasterResolution`, `timeSinceFirePalette`, `maxAge` that are values created externally. Please check if there are any other values that should be moved to arguments.

**Review by DamianRodziewicz — CHANGES_REQUESTED** [review]
One more comment. PTAL

**Comment by mateuszwyszynski** [comment]
Comments addressed. PTAL

File: R/histogramForRaster.R [hunk 4]
```diff
@@ -0,0 +1,64 @@
+#' Histogram for Raster Shiny Module
+#'
+#' @description This creates UI for shiny module creating histogram of data contained in raste
+#'
+#' @param id An ID string that corresponds with the ID used to call the module server function
+#' @param title Optional title for the histogram. Any shiny tag can be used.
+#' @param ... Additional parameters passed to \code{\link[shiny]{box}} tag from \pkg{shiny}.
+#' @param plotParameters A list of parameters passed to \code{\link[shiny]{plotOutput}} function from \pkg{shiny} package.
```
Comment (DamianRodziewicz) [inline R/histogramForRaster.R]: Please set the correct order.

File: R/histogramForRaster.R [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+#' @importFrom shinydashboard box
+#' @importFrom shinycssloaders withSpinner
+#'
+#' @rdname histogramForRaster
+#'
+#' @export
+histogramForRasterUI <- function(id, title = "", plotParameters, ...) {
+  ns <- NS(id)
+
+  plotParameters["outputId"] = NULL
```
Comment (DamianRodziewicz) [inline R/histogramForRaster.R]: Use `<-`

File: R/histogramForRaster.R [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+  plotOutputParameters <- c(ns("histogram"), plotParameters)
+
+  box(title,
+      shinycssloaders::withSpinner(
+        do.call(plotOutput, plotOutputParameters)
+        ),
```
Comment (DamianRodziewicz) [inline R/histogramForRaster.R]: incorrect indentation

File: R/histogramForRaster.R [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+#'
+#' @description
+#'
+#' @param input Shiny server input object
+#' @param output Shiny server output object
+#' @param session Shiny server session object
+#' @param raster Reactive value containing raster
+#' @param scale Number used for scaling heights of histogram bars. When set to 1 (default) histogram bar height represents
+#' amount of raster cells with value from bar interval. If the resolution of raster is known, scale parameter can
+#' be used to transform these heights into the ones representing area covered by cells.
```
Comment (DamianRodziewicz) [inline R/histogramForRaster.R]: Please add more details as to what is the unit of scale, for example what does `scale = 2` mean.

File: R/timeSinceFire.R [hunk 8]
```diff
@@ -20,11 +20,8 @@ timeSinceFireUI <- function(id, rastersNumber) {
                  min = 0, max = (rastersNumber-1)*10, value = 0, step = 10,
                  animate = animationOptions(interval = 2500, loop = FALSE))
     ),
-    box(width = 4, solidHeader = TRUE, collapsible = TRUE,
-        h4(paste("Current time since distribution distribution")),
-        withSpinner(plotOutput(ns("timeSinceFire2Hist"), height = 600))
-    )
-  )
+    histogramForRasterUI(ns("histogram"), title = h4(paste("Current time since distribution distribution")), plotParameters = list(height = 600),
```
Comment (DamianRodziewicz) [inline R/timeSinceFire.R]: Too long line.

File: R/timeSinceFire.R [hunk 9]
```diff
@@ -101,16 +98,19 @@ timeSinceFire <- function(input, output, session, rasters) {
...
-    timeSinceFireHist <- hist(ras1[], plot = FALSE, breaks = Nbreaks)
-    barplot(timeSinceFireHist$counts*prod(rasterResolution)/1e4, xlab = "Time since fire \n(Years)",
-            col = timeSinceFirePalette(1:(maxAge/10)), width = 1, space = 0, ylab = "Area (ha)")
-    axis(1, at = timeSinceFireHist$breaks/10, labels = 0:Nbreaks*10)
+  raster <- reactive(rasterInput()$r)
+
+  breaks <- reactive(ceiling(maxValue(raster())/10))
 
+  addAxisParams <- reactive({
+    numberOfBreaks <- ceiling(maxValue(raster())/10)
```
Comment (DamianRodziewicz) [inline R/timeSinceFire.R]: DRI - `ceiling(maxValue(raster())/10)` is repeated above. Please reuse breaks() reactive value.

**Review by DamianRodziewicz — CHANGES_REQUESTED** [review]
Please address comments, merge develop branch and fix conflicts.

File: R/timeSinceFire.R (hunk 9 above)
Comment (mateuszwyszynski) [reply R/timeSinceFire.R]: I didn't reuse breaks(), because in general breaks might be also a vector (not a number)

Comment (mateuszwyszynski) [reply R/timeSinceFire.R]: Done

**Comment by mateuszwyszynski** [comment]
Comments addressed, PTAL

File: R/histogramForRaster.R [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+#'            See \code{\link[graphics]{barplot}} for reference.
+#'
+#' @importFrom shiny renderPlot
+#' @importFrom raster hist
+#' @importFrom graphics axis barplot
+#'
+#' @rdname histogramForRaster
+#'
+#' @export
+histogramForRaster <- function(input, output, session, raster, scale = 1, histogramBreaks, addAxisParams = NULL,  ...) {
```
Comment (DamianRodziewicz) [inline R/histogramForRaster.R]: Change the order of arguments, so that the ones with default values set are last.

**Review by DamianRodziewicz — APPROVED** [review]
One last change to apply and you can merge.

**Pull request merged**